<!doctype html> 
<html>
<head>
  <title>DLCoal documentation</title>
  <link type="text/css" rel="stylesheet" href="manual.css">
</head>

<body>

<!-- ====================================================================== -->
<div class="titlepage">

<h1>
DLCoal: Modeling gene duplications, losses, and coalescence
</h1>

<h2>
Documentation for the DLCoal software package
</h2>

<h3>
Updated: Dec 21, 2011
</h2>

<b>Author:</b> Matthew D. Rasmussen (<a href="http://mattrasmus.com">mattrasmus.com</a>, <a href="mailto:rasmus@alum.mit.edu">rasmus@alum.mit.edu</a>, <a href="mailto:matt.rasmus@gmail.com">matt.rasmus@gmail.com</a>)<br>
<b>Software website:</b> <a href="http://compbio.mit.edu/dlcoal">http://compbio.mit.edu/dlcoal</a>

<br><br>

<b>citation:</b>
<a href="http://dx.doi.org/10.1101/gr.123901.111">
Matthew D. Rasmussen and Manolis Kellis.
<i>Unified modeling for gene duplication, loss, and coalescence using a locus tree</i>.
Genome Research. 2012.</a>

<br><br>

<hr>

<br><br>

</div>


<!-- ====================================================================== -->
<h2>Introduction</h2>

<p>
The DLCoal software package contains several programs reconciling gene
trees and inferring their gene duplications and losses.  The concepts
and algorithms behind the main
program, <a href="#sec-prog-dlcoal_recon"><tt>dlcoal_recon</tt></a>,
are presented in our <a href="#cite-Rasmussen2012">DLCoal paper</a>.
This documentation describes how to use each of
the <a href="#sec-prog">programs</a> in the software package and defines
the various <a href="#sec-file">file formats</a> that are used.
</p>

<p>
For an example of how to run these programs see
the <a href="../examples/">examples</a> directory
(e.g. <a href="../examples/analyze-sim-flies.sh">analyze-sim-flies.sh</a>).
</p>

<p>
<em>Note: This documentation is currently under development and
will be regularly updated with more information.</em>
</p>


<!-- ===================================================================== -->
<a name="sec-prog"></a>
<h2>Programs</h2>

<!-- ===================================================================== -->
<a name="sec-prog-dlcoal_recon"></a>
<h3>dlcoal_recon</h3>

<p>
<tt>dlcoal_recon</tt> is a program that takes a previously
reconstructed gene tree and a species tree and produces
a <em>reconciliation</em>, a data-structure that describes how the
gene tree "fits inside" the species tree.  From this data structure, one
can infer gene duplications, losses, and incomplete lineage sorting.
Typically, reconciliation is represented as mapping of nodes in a gene tree
to the nodes and edges of the species tree.  However,  DLCoal uses a
<a href="#fig-three-trees">three-tree model</a> (gene tree, locus
tree, and species tree) to represent gene family evolution.  Therefore
the reconciliation data structure is more than just a single mapping.
Instead it is represented as a tuple of four variables: the locus tree
T<sup>L</sup>, the gene-to-locus tree mapping R<sup>G</sup>, the
locus-to-species mapping R<sup>L</sup>, and the set of daughter nodes
&delta;<sup>L</sup>.
</p>

<a name="fig-three-trees"></a>
<div class="figure">

<img src="figures/three-trees.png">

<div class="figcaption">
<span class="figtitle">Figure: DLCoal's three tree model.</span>
DLCoal assumes a generative process for gene family evolution that
follows the following steps.  Given a species tree S with known
topology and divergence times, a top-down dup-loss process generates a
locus tree T<sup>L</sup> , which contains duplication nodes (star) and
each daughter duplicate is indicated by a daughter edge
&delta;<sup>L</sup> (dark red). From the locus tree, the bottom-up
multilocus coalescent (MLC) process generates a gene tree
T<sup>G</sup>. Mappings between the trees represented by R<sup>G</sup>
and R<sup>L</sup> indicate how one tree "fits inside" the other.
</div>
</div>



<h4>Main arguments</h4>

<div class="code">
  -s SPECIES_TREE, --stree=SPECIES_TREE
</div>

<p>
  This specifies the <a href="#sec-file-stree">species tree</a>
  in <a href="http://en.wikipedia.org/wiki/Newick_format">Newick
  format</a>.  The branch lengths should be specified in millions of years
  (myr).
</p>

<div class="code">
  -S GENE_TO_SPECIES_MAP, --smap=GENE_TO_SPECIES_MAP
</div>

<p>
This specifies a <a href="#sec-file-smap">gene to species map</a>.
</p>

<div class="code">
  -n POPULATION_SIZE, --popsize=POPULATION_SIZE
</div>

<p>
This specifies an effective population size N<sub>e</sub> to use for every
branch in the species tree.  Branch-specific population sizes are 
also <a href="#sec-file-stree-params">possible</a>.
</p>

<div class="code">
  -D DUPLICATION_RATE, --duprate=DUPLICATION_RATE
</div>

<p>
  This specifies the rate of a gene duplication (dups/gene/myr) present
  throughout all branches of the species tree.
</p>

<div class="code">
  -L LOSS_RATE, --lossrate=LOSS_RATE
</div>

<p>
  This specifies the rate of a gene loss (losses/gene/myr) present
  throughout all branches of the species tree.
</p>

<div class="code">
  -g GENRATION_TIME, --gentime=GENRATION_TIME
</div>

<p>
  This specifies the generation time (years) present throughout all branches
  of the species tree.  Branch-specific generation times are also 
  <a href="#sec-file-stree-params">possible</a>.
</p>

<div class="code">
  -i ITERATIONS, --iter=ITERATIONS
</div>

<p>
  This specifies the number of search iterations to perform
</p>


<h4>File extensions</h4>

<div class="code">
    -I INPUT_EXT, --inext=INPUT_EXT
</div>

<p>
  This specifies the input file extension (default='') of the gene
  tree.  This extension will be stripped from the input filename in
  order to determine the output file prefix <tt>"PREFIX"</tt>.
</p>

<div class="code">
    -O OUTPUT_EXT, --outext=OUTPUT_EXT
</div>

<p>
  This specifies the output file extension (default='.dlcoal') for all
  output files.  Every output file will have the
  form <tt>"$PREFIX$OUTPUT_EXT.*"</tt>
  (e.g. <tt>"my_input_prefix.dlcoal.locus.recon"</tt>).
</p>



<h4>Miscellaneous</h4>

<div class="code">
    --nprescreen=NUM_PRESCREENS
</div>

<p>
  This specifies the number of prescreening iterations to use in the locus tree
  search (default=50).
</p>

<div class="code">
    --nsamples=NUM_SAMPLES
</div>

<p>
  This specifies the number of samples for dup-loss integration
  (default=100)
</p>

<div class="code">
    --init-locus-tree=TREE_FILE
</div>

<p>
  This specifies the initial locus tree for search.  By default the 
  gene tree topology is used as the initial locus tree.
</p>

<div class="code">
    -x RANDOM_SEED, --seed=RANDOM_SEED
</div>

<p>
  This specifies the seed for the random number generator.  This allows one
  to produce deterministic results.
</p>

<div class="code">
    -l, --log           if given, output debugging log
</div>

<p>
Outputs logging information to a log file <tt>*.log</tt>.
</p>


<div class="code">
  -h, --help
</div>

<p>
Displays a help message
</p>



<a name="sec-prog-dlcoal_recon-output"></a>
<h4>dlcoal_recon output files</h4>

<p>
DLCoal represents the reconciliation as a tuple of four variables: the
locus tree T<sup>L</sup>, the gene-to-locus tree mapping
R<sup>G</sup>, the locus-to-species mapping R<sup>L</sup>, and the set
of daughter nodes &delta;<sup>L</sup>.  Each of these variables is
store in one of the following four output files.
</p>

<ul>
  <li><tt>*.locus.tree</tt>: a locus tree stored in <a href="http://en.wikipedia.org/wiki/Newick_format">Newick format</a>.</li>
  <li><tt>*.coal.recon</tt>: a mapping between the gene tree (*.coal.tree) and the locus tree (*.locus.tree) stored in the <a href="#sec-file-recon">*.recon
file format</a>.</li>
  <li><tt>*.locus.recon</tt>: a mapping between the locus tree (*.locus.tree) and the species tree (*.stree) stored in the <a href="#sec-file-recon">*.recon
file format</a>.</li>
  <li><tt>*.dlcoal.daughters</tt>: a set of daughter nodes stored
    in <a href="#sec-file-daughters">*.daughters file format</a>.</li>
</ul>


<tt>dlcoal_recon</tt> also produces two more output files:

<ul>
  <li><tt>*.coal.tree</tt>: a copy of the gene tree with the 
  internals given names.  This file is created in case the original input 
    gene tree file did not specify internal node names.</li>
  <li><tt>*.log</tt>: a log file that describes the various 
  reconciliations proposed during the reconciliation search.</li>
</ul>


<!-- ===================================================================== -->
<a name="sec-prog-dlcoal_sim"></a>
<h3>dlcoal_sim</h3>

<p>
The <tt>dlcoal_sim</tt> program simulates gene trees using the DLCoal model.
The program takes the following arguments:
</p>


<div class="code">
  -o OUTPUT_DIR, --outputdir=OUTPUT_DIR
</div>

<p>
  This specifies the output directory for the simulation files.
</p>

<div class="code">
  -s SPECIES_TREE, --stree=SPECIES_TREE
</div>

<p>
  This specifies the <a href="#sec-file-stree">species tree</a>
  in <a href="http://en.wikipedia.org/wiki/Newick_format">Newick
  format</a>.  The branch lengths should be specified in millions of years
  (myr).
</p>

<div class="code">
  -n POPULATION_SIZE, --popsize=POPULATION_SIZE
</div>

<p>
  This specifies the effective population size (N<sub>e</sub>) present
  in every branch of the species tree.
</p>

<div class="code">
  -D DUPLICATION_RATE, --duprate=DUPLICATION_RATE
</div>

<p>
  This species the rate of a gene duplication (dups/gene/myr) present in
  every branch of the species tree.
</p>

<div class="code">
  -L LOSS_RATE, --lossrate=LOSS_RATE
</div>

<p>
  This specifies the rate of gene loss (losses/gene/myr) present in every
  branch of the species tree.
</p>

<div class="code">
  -g GENRATION_TIME, --gentime=GENRATION_TIME
</div>

<p>
  This specifies the generation time (years) for every branch in the species 
  tree.
</p>

<div class="code">
  -i NUMBER_OF_SIMULATION, --iter=NUMBER_OF_SIMULATION
</div>

<p>
  This specifies the number of simulationed gene trees to create.
</p>

<div class="code">
  --minsize=MIN_NUMBER_OF_LEAVES
</div>

<p>
  This specifies the minimum number of genes required per family.
  If a family is simulated with fewer genes (due to gene losses), the family
  is rejected and and re-simulated.
</p>

<div class="code">
  -h, --help
</div>

<p>
  If given, show the help message.
</p>


<h4>Output files</h4>

<p>
<tt>dlcoal_sim</tt> will produce several simulated gene families each
stored in their own directory as follows:  
</p>

<ul>
  <li><tt>DATADIR/FAMID/FAMID.coal.tree</tt>: a gene tree stored in
  <a href="http://en.wikipedia.org/wiki/Newick_format">Newick format</a>.</li>
  <li><tt>DATADIR/FAMID/FAMID.coal.recon</tt>:
  a <a href="#sec-file-recon">reconciliation mapping</a> of the gene
  tree to the locus tree.</li>
  <li><tt>DATADIR/FAMID/FAMID.locus.tree</tt>: a locus tree stored in 
    <a href="http://en.wikipedia.org/wiki/Newick_format">Newick format</a>.</li>
  <li><tt>DATADIR/FAMID/FAMID.locus.recon</tt>: 
    a <a href="#sec-file-recon">reconciliation mapping</a> of the locus tree
    to the species tree.</li>
  <li><tt>DATADIR/FAMID/FAMID.daughters</tt>: a set
  of <a href="#sec-file-daughters">daughter nodes</a> in the locus
  tree.</li>
</ul>

<p>
The base directory <tt>DATADIR</tt> is specified by the <tt>-d</tt> option and
each directory <tt>FAMID</tt> is numbered between 0 and value specified by
the <tt>-i</tt> option.
</p>



<!-- ===================================================================== -->
<a name="sec-prog-tree-relations"></a>
<h3>tree-relations</h3>

<p>
The <tt>tree-relations</tt> program reads from standard input a list of
gene tree filenames (one filename per line) and outputs to standard output
a <a href="#sec-file-recon">relation file (*.rel.txt)</a>.
</p>


<div class="code">
  -d, --dir
</div>

<p>
  When given, the directory name containing the gene tree is used as the
  family id (column 1).
</p>

<div class="code">
  -s SPECIES_TREE, --stree=SPECIES_TREE
</div>

<p>
  This specifies the <a href="#sec-file-stree">species tree</a>
  in <a href="http://en.wikipedia.org/wiki/Newick_format">Newick
  format</a>.
</p>


<div class="code">
  -S GENE_TO_SPECIES_MAP, --smap=GENE_TO_SPECIES_MAP
</div>

<p>
  This specifies a <a href="#sec-file-smap">gene to species map</a>.
</p>


<div class="code">
  -T TREE_EXT, --treeext=TREE_EXT
</div>

<p>
  This specifies the file extension of the input gene tree.
</p>

<div class="code">
  -R RECON_EXT, --reconext=RECON_EXT
</div>

<p>
  If given, this specifies the file extension of
  the <a href="#sec-file-recon">reconciliation mapping</a>.  The
  reconciliation mapping file is determined by removing the tree
  file extension and appending the reconciliation file extension.
</p>


<div class="code">
  --no-species-branch   
</div>

<p>
  If given, this suppresses the <tt>species id</tt> column in the 
  duplication columns.
</p>


<div class="code">
  -h, --help
</div>

<p>
  If given, show the help message.
</p>

<!-- ===================================================================== -->
<a name="sec-prog-tree-relations-cmp"></a>
<h3>tree-relations-cmp</h3>

<p>
The <tt>tree-relations-cmp</tt> program takes two 
<a href="#sec-file-rel">relation files</a> and compares them. 
Specifically, the program takes two arguments:
</p>

<div class="code">
  tree-relations-cmp FILE1.rel.txt FILE2.rel.txt
</div>

<p>
<tt>FILE1.rel.txt</tt> should be a <a href="#sec-file-rel">relation
file</a> that describes the <em>true</em> relationships within a gene
family or series of gene families (i.e. the actual relationships
determined from simulated data).  <tt>FILE2.rel.txt</tt> gives the
relations reconstructed by a program like <tt>dlcoal_recon</tt>.  The
output of the program will look something like the following:
</p>

<pre class="code">
dup actual:     46
dup pred:       49
dup sn:         0.95652173913
dup ppv:        0.897959183673
loss actual:    34
loss pred:      43
loss sn:        0.970588235294
loss ppv:       0.767441860465
orth actual:    6856
orth pred:      6758
orth sn:        0.985414235706
orth ppv:       0.999704054454
</pre>


<p>
This output describes the performance of the reconciliation algorithm compared
to the truth.  Performance is measured for gene duplications (dup), gene 
losses (loss), and orthologs (orth).  For each type of event or relationship,
both the actual and predicted (pred) counts are reported.  In addition,
the sensitivity (sn)
and the positive predictive value (ppv; also known as precision) are given.
</p>

<p>
<em>
Note: sensitivity = true positives / (true positives + false negatives).
precision = true positives / (true positives + false positives).
</em>
</p>


<!-- ====================================================================== -->
<a name="sec-prepare"></a>
<h2>Preparing your data set</h2>


<p>
<b>Restrictions on gene IDs and species IDs.</b>  Due to the
file formats that DLCoal uses, there as several restrictions on what IDs
are allowed.  Many of these restrictions are common for other similar
phylogenetic software.  The safest IDs follow these restrictions:
</p>

<ol>
<li>the first and last characters of the ID are <tt>a-z A-Z 0-9 _ - .</tt></li>
<li>the middle characters can be <tt>a-z A-Z 0-9 _ - .</tt> or the space character <tt>' '</tt>.</li>
<li>the ID should not be purely numerical characters 0-9</li>
<li>the ID should be unique within a gene tree or within a species tree</li>
</ol>

<p>
Space characters are discouraged from gene IDs and species IDs since
they will probably cause problems with other bioinfomatic software
that you may use (although DLCoal can handle spaces).  Characters
such as parentheses "(" ")" and colons ":" are not allowed
because the <a href="#sec-file-stree">newick file format</a> uses these
characters for describing the structure of the tree.
</p>

<p>
It is also easier to use gene IDs that have a prefix or suffix that indicates
the species ID.  For example "human_HOXC5" is a human gene.  This is not
a requirement, but it does make preparing a <a href="#sec-file-smap">gene to 
species mapping file (*.smap)</a> easier.
</p>




<!-- ====================================================================== -->
<a name="sec-file"></a>
<h2>File formats</h2>


<!-- ====================================================================== -->
<a name="sec-file-stree"></a>
<h3>Species tree file format (*.stree)</h3>


<p>
Species trees should be specified using the 
<a href="http://en.wikipedia.org/wiki/Newick_format">Newick file
format</a>.  Beyond the newick format, DLCoal has only a few
additional requirements.  First, the species names given in the
species tree should match those given in
the <a href="#sec-file-smap">SMAP file</a>.  Second, the branch
lengths of the species tree should be expressed in units of time
(see <a href="#fig-stree">figure</a>).  Any unit of time can be used
(e.g. millions of years, generations, relative units, etc).  The only
requirement is that the duplication and loss rates are also expressed
in compatible units.  Therefore, if branch lengths are in <em>
millions of years</em>, the duplication rate (specified
by <tt>dlcoal_recon</tt>'s "-D" option) should be in units of
duplications/gene/<em>million years</em>.
</p>

<p>
<b>Naming ancestral nodes.</b>
DLCoal also supports naming ancestral nodes in the species tree using
the newick format.  For example, the parental node of <tt>human</tt> and 
<tt>chimp</tt> can be named <tt>primate</tt> using the following syntax:
</p>

<div class="code">
((human:5,chimp:5)primate:70,mouse:75)mammal;
</div>

<p>
If ancestral nodes are named, they will be used in the output of the 
<a href="#sec-file-recon">reconciliation mapping</a>.
</p>

<a name="fig-stree"></a>
<div class="figure">

<pre class="code">
(((((((scer:7.061760,
       spar:7.061760
      )n7:4.999680,
      smik:12.061440
     )n6:5.970600,
     sbay:18.032040
    )n5:52.682400,
    cgla:70.714260
   )n4:7.220700,
   scas:77.934960
  )n3:23.181480,
  (
   (
    agos:78.553260,
    klac:78.553260
   )n9:10.434960,
   kwal:88.988220
  )n8:12.128400
 )n2:78.883560,
 (
  (
   (
    calb:41.275620,
    ctro:41.275980
   )n12:29.632860,
   (
    cpar:52.323120,
    lelo:52.323120
   )n13:18.585720
  )n11:31.149540,
  (
   (
    cgui:75.615840,
    dhan:75.615840
   )n15:14.006880,
   clus:89.622720
  )n14:12.435660
 )n10:77.941620
)n1;
</pre>


<img src="figures/fungi-stree.png">

<div class="figcaption">
<span class="figtitle">Figure: Example *.stree file and corresponding
 tree.</span>  This file (top) specifies the species tree (bottom)
 using the newick file format.  Branch lengths should be expressed in
 units of time (e.g. millions of years).  Ancestral nodes can also
 optionally be named (the names "n1", "n2", etc are used in this
 example).
</div>
</div>


<a name="sec-file-stree-params"></a>
<h4>Branch-specific parameters</h4>

<p>
Several parameters for DLCoal can be specified in a branch-specific way.
Currently, the effective population size and generation time parameters
can be configured in this way.  This is done by using the comment section
within the species tree newick file.  The comment section has the format:
</p>

<div class="code">
[&&NHX:key1=val1:key2=val2]
</div>

<p>
where <tt>&&NHX:</tt> must appear first followed by a series of key-value
pairs each of which are separated by colons ":".  One comment section
can appear after every node in the newick file.  For example, the following
species tree has comments after every node:
</p>

<pre class="code">
(
 scer:610.916819[&&NHX:g=0.9:pop=10000000.0],
 (
  (
   cjaponica:66.116017[&&NHX:g=0.166666666667:pop=50000],
   (
    celegans:18.0[&&NHX:g=0.166666666667:pop=50000],
    (
     cbrenneri:15.79308[&&NHX:g=0.166666666667:pop=50000],
     (
      cremanei:13.973676[&&NHX:g=0.166666666667:pop=50000],
      cbriggsae:13.973676[&&NHX:g=0.166666666667:pop=50000]
     ):1.819404[&&NHX:g=0.166666666667:pop=50000]
    ):2.20692[&&NHX:g=0.166666666667:pop=50000]
   ):48.116017[&&NHX:g=0.166666666667:pop=50000]
  ):386.30503[&&NHX:g=0.166666666667:pop=50000],
  (
   (
    (
     dgri:38.152242[&&NHX:g=0.1:pop=10000000.0],
     (
      dmoj:26.13958[&&NHX:g=0.1:pop=10000000.0],
      dvir:26.13958[&&NHX:g=0.1:pop=10000000.0]
     ):12.012662[&&NHX:g=0.1:pop=10000000.0]
    ):23.847758[&&NHX:g=0.1:pop=10000000.0],
    (
     dwil:54.260773[&&NHX:g=0.1:pop=10000000.0],
     (
      (
       dpse:2.112373[&&NHX:g=0.1:pop=10000000.0],
       dper:2.112373[&&NHX:g=0.1:pop=10000000.0]
      ):40.180327[&&NHX:g=0.1:pop=10000000.0],
      (
       dana:30.860312[&&NHX:g=0.1:pop=10000000.0],
       (
        (
         dyak:7.966764[&&NHX:g=0.1:pop=10000000.0],
         dere:7.966764[&&NHX:g=0.1:pop=10000000.0]
        ):3.264526[&&NHX:g=0.1:pop=10000000.0],
        (
         (
          dsim:2.743866[&&NHX:g=0.1:pop=10000000.0],
          dsec:2.743866[&&NHX:g=0.1:pop=10000000.0]
         ):2.968142[&&NHX:g=0.1:pop=10000000.0],
         dmel:5.712008[&&NHX:g=0.1:pop=10000000.0]
        ):5.519282[&&NHX:g=0.1:pop=10000000.0]
       ):19.629022[&&NHX:g=0.1:pop=10000000.0]
      ):11.432388[&&NHX:g=0.1:pop=10000000.0]
     ):11.968073[&&NHX:g=0.1:pop=10000000.0]
    ):7.739227[&&NHX:g=0.1:pop=10000000.0]
   ):217.014905[&&NHX:g=0.1:pop=10000000.0],
   (
    mouse:75.0[&&NHX:g=0.5:pop=460000],
    human:75.0[&&NHX:g=25:pop=10400]
   ):204.014905[&&NHX:g=12.75:pop=235200.0]
  ):173.406143[&&NHX:g=6.425:pop=5117600.0]
 ):158.495771[&&NHX:g=3.29583333333:pop=2583800.0]
):0[&&NHX:g=2.09791666667:pop=6291900.0];
</pre>

<p>
The following keys are currently supported by 
<a href="#sec-prog-dlcoal_recon"><tt>dlcoal_recon</tt></a>:
</p>

<ul>
  <li><tt>g</tt>: generation time in years.</li>
  <li><tt>pop</tt>: effective population size.</li>
</ul>




<!-- ====================================================================== -->
<a name="sec-file-smap"></a>
<h3>Gene to species name mapping file format (*.smap)</h3>


<p>
DLCoal uses a special file format (*.smap) to specify which
genes belong to which species. Each line contains two tab-delimited
fields: 
</p>

<ol>
  <li> pattern matching a gene ID</li>
  <li> species ID</li>
</ol>

<p>
Only 3 types of gene ID patterns are supported.  The pattern can
either be an exact matching string, a prefix (denoted <tt>"text*"</tt>),
or a suffix (denoted <tt>"*text"</tt>).  The <tt>"*"</tt> is the only special
wildcard character.
</p>

<p>
The species ID should be the same as those used in the species tree.
All patterns and IDs are case-sensitive.
</p>

<p>
When mapping a gene ID to a species ID all exact matches are processed
first.  If no exact match is found, the patterns are then processed in
the same order as they appear in the file until a match is found. For
example in the SMAP file given <a href="#fig-smap">below</a>, the gene
ID <tt>"YALI123"</tt> should match the species <tt>"ylip"</tt>,
instead of <tt>"scer"</tt>, because the pattern <tt>"YALI*"</tt> occurs before 
<tt>"Y*"</tt>.
</p>

<a name="fig-smap"></a>
<div class="figure">

<pre class="code">
A*      agos
orf19*  calb
CDUG_*  cdub
CAGL*   cgla
IPF_*   cgla
CGUG_*  cgui
sbay_*  sbay
scas_*  scas
smik_*  smik
spar_*  spar
SCP*    spom
YALI*   ylip
Y*      scer
Q*      scer
</pre>

<div class="figcaption">
<span class="figtitle">Figure: Example *.smap file.</span> This file
specifies how to map gene names to their corresponding species.  The
first column indicates a gene name pattern (in this case a prefix) and
the second column specifies a species name.  Note: this example only
gives a partial list of the species in
the <a href="#fig-stree">species tree above</a>.
</div>
</div>

<!-- ====================================================================== -->
<a name="sec-file-recon"></a>
<h3>Reconciliation file format (*.recon)</h3>


<p>
The <a href="#sec-prog-dlcoal_recon"><tt>dlcoal_recon</tt></a> and
<a href="#sec-prof-mpr"><tt>mpr</tt></a> programs both produce reconciliation 
mappings which are stored in the reconciliation file format (*.recon)
(see <a href="#fig-recon">reconcile example</a>).  The reconciliation file
format is tab-delimited, where each line has three fields:
</p>

<ol>
  <li>gene node ID.</li>
  <li>species node ID.</li>
  <li>event (one of the following: <tt>"gene"</tt>, <tt>"spec"</tt>, <tt>"dup"</tt>, <tt>"none"</tt>)</li>
</ol>

<p>
Each line specifies the mapping of one node in the gene tree (field 1) to one 
node or branch in the species tree (field 2).  Branches are indicate using
the node ID directly below it (i.e. the younger of the two incident nodes).
The lines can be given in any order.
</p>

<p>
If the gene node is a leaf, it will map to a leaf in the species tree
and the event field will contain the event <tt>"gene"</tt>.  All
internal nodes of the gene tree are marked either as speciations
(event <tt>"spec"</tt>) or duplications (event <tt>"dup"</tt>).
Specaition nodes map directly to the indicated species node, and
duplication nodes map to the indication species branch.  The time of
the duplication along the species branch is not indicated in this file
format nor is it inferred by DLCoal.  If the event is unknown or represents
coalescence the <tt>"none"</tt> keyword is used in column 3.
</p>

<p>
If gene IDs are not given to the ancestral nodes of a gene tree or
species tree, DLCoal will by default name them with <tt>"nXXX"</tt>
where <tt>XXX</tt> is the preorder traversal of the internal nodes.
</p>


<a name="fig-recon"></a>
<div class="figure">

<pre class="code">
KLLA0C08239g  klac  gene  
ADL072C       agos  gene  
kwal_5828     kwal  gene  
CAGL0J02970g  cgla  gene  
scas_g715.48  scas  gene  
smik_6662     smik  gene  
sbay_7039     sbay  gene  
smik_6659     smik  gene  
sbay_7037     sbay  gene  
YER061C       scer  gene  
spar_6281     spar  gene  
n10           n5    spec  
n9            n7    spec  
n8            n6    spec  
n7            n5    spec  
n6            n5    dup   
n5            n3    spec  
n4            n3    spec   
n3            n9    spec  
n2            n8    spec  
n1            n2    spec  
</pre>

<img src="figures/gene-tree.png">

<div class="figcaption">
<span class="figtitle">Figure: Example *.recon file.</span> The
reconciliation file format (top) specifies how all the nodes in a
gene tree (bottom) map to the nodes and branches in the species tree
(see <a href="#fig-stree">species tree example</a>).  Notice that gene
node "n6" (red dot) represents a duplication event along species
branch "n5" (shown in the <a href="#fig-stree">species tree
example</a>).  The gene tree and species tree have their own name
space (e.g. "n5" in the gene tree is not the same as "n5" in the species
tree).
</div>
</div>


<h4>Special notes about DLCoal's use of reconciliation mappings</h4>

<p>
As we discuss in the <a href="#cite-Rasmussen2012">DLCoal paper</a>,
DLCoal uses a <a href="#fig-three-trees">three-tree model</a> of gene
family evolution (gene tree, locus tree, species tree).
<a href="#sec-prog-dlcoal_recon"><tt>dlcoal_recon</tt></a> constructs
mappings between the gene tree and locus tree as well as between the 
locus tree and species tree.  Both of these mappings are stored in
the reconciliation file format (*.recon).  For more information on how to
interpret these mappings see 
<a href="#sec-prog-dlcoal_recon-output">dlcoal_recon output files</a>.
</p>


<!-- ====================================================================== -->
<a name="sec-file-daughters"></a>
<h3>Daughter nodes file format (*.daughters)</h3>

<p>
Part of DLCoal's reconciliation representation is the set of daughter
nodes &delta;<sup>L</sup> in the locus tree T<sup>L</sup>.  This
set is stored in a very simple file format (*.daughters).  Each line of a
a *.daughters file contains one locus node id that belongs to the daughter
set.  The node ids can be specified in any order.
</p>


<!-- ====================================================================== -->
<a name="sec-file-rel"></a>
<h3>Gene relationship file format (*.rel.txt)</h3>

<p>
After reconciling a gene tree,
the <a href="#sec-prog-tree-relations"><tt>tree-relations</tt></a>
program can be used to determine the relationship (i.e. orthology,
paralogy, etc) between genes in a gene family.  These relationships
are stored in a special file format called a <em>relation file
(*.rel.txt)</em>.  The file format is tab-delimited and has the following
columns:
</p>

<ol>
<li>famid</li>
<li>relationship type (one of "gene", "spec", "dup", "loss")</li>
<li>relationship type specific data, see below</li>
<li>all additional columns are also relationship type specific</li>
</ol>

<p>
Each row specifies one "relationship" between sets of genes. The kind
of relationship is indicated by column 2 (relationship type) and are
defined as follows.</p>

<h4>gene relationship</h4>

<p>
Relationship "gene" has the columns:
</p>

<ol>
<li value="3">gene id</li>
</ol>

<p>
This indicates that the gene with <tt>gene id</tt> belongs to
family <tt>famid</tt>.
</p>

<h4>spec (speciation/ortholog) relationship</h4>
<p>
Relationship "spec" (speciation) has the columns:
</p>

<ol>
  <li value="3">gene1,gene2,gene3,...  genes on the left side of a speciation node</li>
  <li>geneA,geneB,geneC,...  genes on the right side of a speciation node</li>
  <li>species id</li>
</ol>

<p>
Two genes are defined to be <em>orthologs</em> if their most recent common
ancestor (MRCA) is a speciation node in the gene tree. The "spec"
relationship specifies a speciation node that reconciles to the
species <tt>species id</tt> and it also indicates all the descendants
in its left (column 3) and right (column 4) subtrees. Therefore, every
pair of genes from column 3 and column 4 is an ortholog pair. The
ortholog pairs for the above example are:
</p>

<pre>
gene1 geneA
gene1 geneB
gene1 geneC
gene2 geneA
gene2 geneB
gene2 geneC
gene3 geneA
gene3 geneB
gene3 geneC
</pre>


<h4>dup (duplication/paralog) relationship</h4>

<p>
The relationship "dup" (duplication) has the columns:
</p>

<ol>
  <li value="3">gene1,gene2,gene3,...  genes on the left side of a duplication node</li>
  <li>geneA,geneB,geneC,...  genes on the right side of a duplication node</li>
  <li>species id</li>
</ol>

<p>
The "dup" relationship is defined in a veyr similar way as the "spec"
relationship.  Two genes are defined to be <em>paralogs</em> if their most
recent common ancestor (MRCA) is a duplication node in the gene
tree.  The "dup" relationship specifies a duplication node that
reconciles to the species branch <tt>species id</tt> and it also indicates
all the descendants in its left (column 3) and right (column 4)
subtrees. Therefore, every pair of genes from column 3 and column 4 is
a paralog pair.
</p>


<h4>loss relationship</h4>

<p>
The relationship "loss" has the columns:
</p>

<ol>
  <li value="3">gene node id</li>
  <li>species branch id</li>
</ol>

<p>
The loss relationship indicates that a gene was lost in this gene
family (<tt>famid</tt>).  Column 3 gives the node id in the gene tree where
an addition branch was expected to branch off, but is not present because it
has been lost.  Column 4 gives the species branch in which the gene loss
occurred.
</p>


<h4>Complete example</h4>

<p>
For example of a complete relation file for a single gene family, see
the <a href="#fig-rel">figure</a> below.  The relation file format is
also designed to store the relationships of multiple gene families at
the same time.  Simply use a a unique <tt>famid</tt> in column 1 for each 
gene family.
</p>

<a name="fig-rel"></a>
<div class="figure">

<pre class="code">
100.yeast.tree	gene	YER061C
100.yeast.tree	gene	spar_6281
100.yeast.tree	gene	smik_6662
100.yeast.tree	gene	sbay_7039
100.yeast.tree	gene	smik_6659
100.yeast.tree	gene	sbay_7037
100.yeast.tree	gene	CAGL0J02970g
100.yeast.tree	gene	scas_g715.48
100.yeast.tree	gene	kwal_5828
100.yeast.tree	gene	ADL072C
100.yeast.tree	gene	KLLA0C08239g
100.yeast.tree	spec	sbay_7037	smik_6659	n5
100.yeast.tree	spec	YER061C,spar_6281	smik_6662	n6
100.yeast.tree	spec	YER061C	spar_6281	n7
100.yeast.tree	dup	YER061C,sbay_7039,smik_6662,spar_6281	sbay_7037,smik_6659	5
100.yeast.tree	spec	ADL072C,KLLA0C08239g,kwal_5828	CAGL0J02970g,YER061C,sbay_7037,sbay_7039,scas_g715.48,smik_6659,smik_6662,spar_6281	n2
100.yeast.tree	spec	ADL072C,KLLA0C08239g	kwal_5828	n8
100.yeast.tree	spec	ADL072C	KLLA0C08239g	n9
100.yeast.tree	spec	CAGL0J02970g,YER061C,sbay_7037,sbay_7039,smik_6659,smik_6662,spar_6281	scas_g715.48	n3
100.yeast.tree	spec	CAGL0J02970g	YER061C,sbay_7037,sbay_7039,smik_6659,smik_6662,spar_6281	n4
100.yeast.tree	spec	YER061C,smik_6662,spar_6281	sbay_7039	n5
100.yeast.tree	loss	smik_6659	n7
</pre>


<img src="figures/gene-tree.png">

<div class="figcaption">
<span class="figtitle">Figure: Example relation file (*.rel.txt)</span>
The relation file (top) describes on each line a different gene relationship 
that exists in the gene family (below).</span>
</div>
</div>


<h4>Special note about gene id ordering</h4>

<p>
Often times users will want to compare different relation files
generated for the same gene families.  For example, one might want to
compare the true relationships from a simulated data set and the
reconstructed relationships using DLCoalRecon or another method.  This
can be done using
the <a href="#sec-prog-tree-relations-cmp"><tt>tree-relations-cmp</tt></a>
program, which can directly compare two relation files.
</p>

<p>
In order to facilitate comparison of relation files the 
<a href="#sec-prog-tree-relations"><tt>tree-relations</tt></a> program
outputs gene ids for the "spec" and "dup" relationships in a
deterministic order.  Specifically, gene ids are sorted in
alphabetical order within columns 3 and 4 and the subtree with the
alphabetically least gene id is given in column 3.  Therefore, we
can define that a particular duplication is present within two different
relation files if we have the same values for columns 3 and
4 (and optionally 5) for a "dup" relation in both relation files.
</p>


<!-- ====================================================================== -->
<h2>References</h2>

<ul>
  <li><a name="cite-Rasmussen2012"></a>Matthew D. Rasmussen and Manolis Kellis. A unified model for gene duplication, loss, and coalescence using a locus tree. in prep. 2012.</li>
</ul>

</body>
</html>
